packages/flutter/lib/src/widgets/editable_text.dart

```diff
@@ -2236,6 +2235,20 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
       }
     }
 
+    if (_value.selection != value.selection) {
+      try {
+        widget.onSelectionChanged?.call(value.selection, null);
```

Comment:
We don't seem to have a selection change cause when it is caused by selection overlay, I passing null instead. Should we create a new cause? This method will be called when user drag selection handles or tap on tool bar select all. I don't think we have a way to distinguish them.

Reply:
Would it be possible to call onSelectionChanged somewhere higher up in the call stack where you can distinguish drag/select-all?  That might also help work around the duplicate call to onSelectionChanged that @LongCatIsLooong mentioned above.

Reply:
I think we should create an issue to track the new cause creation, add implement it if lots of people are interested in it.

LGTM, once you update the docs of `onSelectionChanged` to make it clear in which cases the `cause` will be null.

Reply:
The only place we can tell what causes the selection change is in the selection overlay because it just set textEditingValue for all the changes happens in selection overlay. I am struggling to find a clean API to be able to move the logic into selection overlay.

The problem that @LongCatIsLooong  mentioned is a bug in render editable i think.

So handleMovement does two thing in this order
1. set textEditingValue
2. call onSelectionChange

This is wrong because the selection would have been updated in step 1, so step 2 is calling selection change with the same selection.

The correct order should be as following
1. call onSelectionChange
2. set textEditingValue

I am not sure whether 2 is even needed.

Reply:
So I decide to leave at the same place. If we want to move it higher up, we will have to add API to TextSelectionDelegate to update the selection, and makes selection overlay to call both set textEditingValue and  the new update selection API which is quite ugly. I am still open to suggestion though.

Reply:
That sounds good to me.  I think getting this right will require a bigger refactor, which hopefully is upcoming.  No need to hack something now.